tests/integration/tls/test_tls.py

```diff
@@ -154,3 +157,72 @@ async def test_tls_renewal(ops_test: OpsTest) -> None:
         updated_certs["http_certificates_list"][0]["not_before"]
         > current_certs["http_certificates_list"][0]["not_before"]
     )
+
+
+@pytest.mark.runner(["self-hosted", "linux", "X64", "jammy", "large"])
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_tls_expiration(ops_test: OpsTest) -> None:
+    """Test that expiring TLS certificates are renewed."""
+    # before we can run this test, need to clean up and deploy with different config
+    await ops_test.model.remove_application(APP_NAME, block_until_done=True)
+    await ops_test.model.remove_application(TLS_CERTIFICATES_APP_NAME, block_until_done=True)
+
+    # Deploy TLS Certificates operator
+    config = {"ca-common-name": "CN_CA", "certificate-validity": "1"}
+    await ops_test.model.deploy(TLS_CERTIFICATES_APP_NAME, channel="stable", config=config)
+    await wait_until(ops_test, apps=[TLS_CERTIFICATES_APP_NAME], apps_statuses=["active"])
+
+    # Deploy Opensearch operator
+    await ops_test.model.set_config(MODEL_CONFIG)
+    my_charm = await ops_test.build_charm(".")
+    await ops_test.model.deploy(
+        my_charm,
+        num_units=1,
+        series=SERIES,
+    )
+
+    await wait_until(
+        ops_test,
+        apps=[APP_NAME],
+        units_statuses=["blocked"],
+        wait_for_exact_units=1,
+    )
+
+    # Now apply a hack to make the certificate secrets expire in short time
+    # set the secret expiry to a fixed timedelta of 300 seconds to give time to start initially
+    # this happens on the tls_certificates lib and we apply the patch via sed-command
+    search_expression = "expire=self._get_next_secret_expiry_time\\(certificate\\)"
+    replace_expression = "expire=timedelta\\(seconds=180\\)"
+
+    unit_id = get_application_unit_ids(ops_test, APP_NAME)[0]
+    unit_ip = await get_leader_unit_ip(ops_test)
+    lib_file = f"/var/lib/juju/agents/unit-opensearch-{unit_id}/charm/lib/charms/tls_certificates_interface/v3/tls_certificates.py"
+    cmd = f"juju ssh {APP_NAME}/{unit_id} sudo sed -i 's/{search_expression}/{replace_expression}/g' {lib_file}"
+    subprocess.run(cmd, shell=True)
+
+    # Relate OpenSearch to TLS and wait until all is settled
+    await ops_test.model.integrate(APP_NAME, TLS_CERTIFICATES_APP_NAME)
+    assert await is_up(ops_test, unit_ip), "OpenSearch service hasn't started."
+    logger.info("Waiting for certificates to expire.")
+
+    # wait for the certs API to be ready
+    time.sleep(30)
+
+    # now start with the actual test
+    # first get the currently used certs
+    current_certs = await get_loaded_tls_certificates(ops_test, unit_ip)
+
+    # now wait for the expiration period to pass by (and a bit longer for things to settle)
+    # we can't use `wait_until` here because the unit might not be idle in the meantime
+    time.sleep(240)
```

Comment:
Can't we wait until the unit becomes idle?
Or for any other condition that has to unavoidably occur?

Again it would be resulting in more stable pipelines...
We had a LOT of trouble in the past with unstable Opensearch pipelines...
I think we should be more strict on preserving current reliable quality and avoid any shadow of flakyness...

Reply:
Same remark as above: No, unfortunately not. As the certificates will be reloaded during that waiting period, the unit would not be active/idle anymore and the wait time would start again. Instead we have a fixed time after that we expect the certs to be renewed, that's why I chose `sleep` here.

Reply:
In this case, a constant was added to signal this sleep is not just some arbitrary time.